pkg/segment/segexecution.go

```diff
@@ -93,136 +101,120 @@ func HelperQueryArithmeticAndLogical(queryOps []structs.QueryArithmetic, resMap
-						if swapped {
-							isLte = valueLHS >= valueRHS
-						}
-						if isLte {
-							finalResult[groupID][timestamp] = valueLHS
-						}
-					case utils.LetEquals:
-						if valueLHS == valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
-						}
-					case utils.LetNotEquals:
-						if valueLHS != valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+					putils.SetFinalResult(queryOp, finalResult, groupID, timestamp, valueLHS, valueRHS, swapped)
+				}
+			}
+
+		} else {
+			// Since each grpID is unique and contains label set information, we can map lGrpID to labelSet and labelSet to rGrpID.
+			// This way, we can quickly find the corresponding rGrpID for a given lGrpID in the other vector. If there is no corresponding result, it means there are no matching labels between the two vectors.
+			idToMatchingLabelSet := make(map[string]string)
+			matchingLabelValTorightGroupID := make(map[string]string)
+			hasVectorMatchingOp := queryOp.VectorMatching != nil && len(queryOp.VectorMatching.MatchingLabels) > 0
+			if hasVectorMatchingOp {
+				// Place the vector with higher cardinality on the left side.
+				if queryOp.VectorMatching.Card == structs.CardOneToMany {
+					swapped = true
+					resultLHS = resMap[queryOp.RHS]
+					resultRHS = resMap[queryOp.LHS]
+				}
+
+				matchingLabelsComb := make(map[string]struct{}, 0)
+				for lGroupID := range resultLHS.Results {
+					matchingLabelValStr := putils.ExtractMatchingLabelSet(lGroupID, queryOp.VectorMatching.MatchingLabels, queryOp.VectorMatching.On)
+					// If the left vector is the 'One' (vector with lower cardinality), it should not have repeated MatchingLabels.
+					// E.g.: (metric1{type="compact"} on (color,type) group_right metric2), the value combinations of (color,type) must be unique in the metric1
+					if queryOp.VectorMatching.Card == structs.CardOneToOne {
+						_, exists := matchingLabelsComb[matchingLabelValStr]
+						if exists {
+							return nil, fmt.Errorf("HelperQueryArithmeticAndLogical: many-to-many matching not allowed: matching labels must be unique on one side")
 						}
+						matchingLabelsComb[matchingLabelValStr] = struct{}{}
+					}
+					idToMatchingLabelSet[lGroupID] = matchingLabelValStr
+				}
+
+				matchingLabelsComb = make(map[string]struct{}, 0)
+				for rGroupID := range resultRHS.Results {
+					matchingLabelValStr := putils.ExtractMatchingLabelSet(rGroupID, queryOp.VectorMatching.MatchingLabels, queryOp.VectorMatching.On)
+					// Right vector is always the 'One' (vector with lower cardinality), it should not have repeated MatchingLabels.
+					_, exists := matchingLabelsComb[matchingLabelValStr]
+					if exists {
+						return nil, fmt.Errorf("HelperQueryArithmeticAndLogical: many-to-many matching not allowed: matching labels must be unique on one side")
 					}
+					matchingLabelsComb[matchingLabelValStr] = struct{}{}
+
+					matchingLabelValTorightGroupID[matchingLabelValStr] = rGroupID
 				}
```

Comment:
I'm confused. Line 140 says `Right vector is always the 'One'`, but line 125 says `If the left vector is the 'One' ...`. Can they both be the One? What is the `One`?

Reply:
It describes the cardinality relationship of two Vectors in a binary operation. So it has `One-to-one vector matches` and `Many-to-one and one-to-many vector matches`

Reply:
Ok, so on line 143 we'd change that later if we add many-to-many operators? And for one-to-many and many-to-one we'll swap left and right if necessary so that the right is the `one` in those cases?

Reply:
Yes, right will also be the `one` in this case. i will add a comment above to mark that. If we add `many-to-many operators` in the future, we should change line 143.